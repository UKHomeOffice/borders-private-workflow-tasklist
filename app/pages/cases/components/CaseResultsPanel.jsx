import React from 'react';
import {withRouter} from "react-router";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import PropTypes from "prop-types";
import _ from "lodash";
import {getCaseByKey} from '../actions';
import {caseDetails, loadingCaseDetails, loadingNextSearchResults} from "../selectors";
import withLog from "../../../core/error/component/withLog";
import CaseDetailsPanel from "./CaseDetailsPanel";
import DataSpinner from "../../../core/components/DataSpinner";

class CaseResultsPanel extends React.Component {

    componentDidMount() {
        if (this.props.businessKey) {
            this.props.getCaseByKey(this.props.businessKey);
        }
    }

    render() {
        const {
            searching, caseSearchResults,
            businessKeyQuery, loadingCaseDetails, caseDetails,
            loadingNextSearchResults
        } = this.props;

        if (searching) {
            return <h4 className="govuk-heading-s">Searching cases with reference {businessKeyQuery}...</h4>
        }
        if (!caseSearchResults) {
            return <div />
        }
        const hasMoreData = _.has(caseSearchResults._links, 'next');

        if (!caseSearchResults._embedded) {
            caseSearchResults._embedded = {
                cases: []
            }
        }
        const businessKeys = caseSearchResults._embedded.cases.map(c => {
            return (
              <li key={c.businessKey}><a
                className="govuk-link"
                href=""
                onClick={event => {
                event.preventDefault();
                if (this.props.businessKey !== c.businessKey) {
                    this.props.getCaseByKey(c.businessKey);
                }
            }}
              >{c.businessKey}
              </a>
              </li>
)
        });
        return (
          <div className="govuk-grid-row">
            <div className="govuk-grid-column-one-quarter">
              <h3 className="govuk-heading-m">Search results</h3>
              {caseSearchResults ? (
                <React.Fragment>
                  <span className="govuk-caption-m">Number of cases found</span>
                  <h3 className="govuk-heading-m">{caseSearchResults.page.totalElements}</h3>

                  {caseSearchResults.page.totalElements > 0 ? (
                    <ul className="govuk-list">
                      {businessKeys}
                    </ul>
) : null}
                  {hasMoreData ? (
                    <button
                      className="govuk-button"
                      disabled={loadingNextSearchResults}
                      onClick={event => {
                                        event.preventDefault();
                                        this.props.loadNext();
                                    }}
                    >{loadingNextSearchResults ? 'Loading more' : 'Load more'}
                    </button>
                          ) : null}
                </React.Fragment>
                  ) : null}
            </div>
            <div className="govuk-grid-column-three-quarters">
              {
                    loadingCaseDetails ? (
                      <div style={{justifyContent: 'center', paddingTop: '20px'}}>
                        <DataSpinner
                          message="Loading case details"
                        />
                      </div>
) :
                        (!caseDetails ? null : <CaseDetailsPanel {...{caseDetails}} />)
                }

            </div>
          </div>
)
    }
}

CaseResultsPanel.propTypes = {
    getCaseByKey: PropTypes.func.isRequired,
    loadNext: PropTypes.func,
    loadingNextSearchResults: PropTypes.bool,
    loadingCaseDetails: PropTypes.bool,
    businessKey: PropTypes.string,
    caseDetails: PropTypes.shape({
        businessKey: PropTypes.string,
        processInstances: PropTypes.arrayOf(PropTypes.object)
    })
};

const mapDispatchToProps = dispatch => bindActionCreators({getCaseByKey}, dispatch);

export default withRouter(connect(state => {
    return {
        kc: state.keycloak,
        appConfig: state.appConfig,
        caseDetails: caseDetails(state),
        loadingCaseDetails: loadingCaseDetails(state),
        loadingNextSearchResults: loadingNextSearchResults(state)

    }
}, mapDispatchToProps)(withLog(CaseResultsPanel)));
