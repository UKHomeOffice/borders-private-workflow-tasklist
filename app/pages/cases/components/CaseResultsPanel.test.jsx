import React from "react";
import CaseResultsPanel from "./CaseResultsPanel";
import {Router} from "react-router";
import {Provider} from "react-redux";
import {createMemoryHistory} from "history";
import configureStore from "redux-mock-store";
import {Map} from "immutable";

describe('CaseResultsPanel', () => {
    it('renders without crashing', () => {
        const wrapper = shallow(<CaseResultsPanel/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('renders searching text', async () => {
        const history = createMemoryHistory('/case');
        const props = {
            searching: true,
            businessKeyQuery: 'BF-'
        };
        const store = configureStore()({
            'case-page': Map({
            }),
            keycloak: props.kc,
            appConfig: props.appConfig
        });

        const wrapper = mount(<Router history={history}>
            <Provider store={store}>
                <CaseResultsPanel {...store} {...props}/>
            </Provider></Router>);

        expect(wrapper.find('h4').exists()).toEqual(true);
        expect(wrapper.find('h4').text()).toEqual('Searching cases with reference BF-...')
    });

    it('returns empty div if no results', async() => {
        const history = createMemoryHistory('/case');
        const props = {
            searching: false,
            businessKeyQuery: 'BF-',
            caseSearchResults: null
        };
        const store = configureStore()({
            'case-page': Map({
            }),
            keycloak: props.kc,
            appConfig: props.appConfig
        });

        const wrapper = mount(<Router history={history}>
            <Provider store={store}>
                <CaseResultsPanel {...store} {...props}/>
            </Provider></Router>);

        expect(wrapper.find('div').exists()).toEqual(true);
        expect(wrapper.find('div').children().length).toEqual(0)
    });
    it('renders search result keys', async() => {
        const history = createMemoryHistory('/case');
        const props = {
            searching: false,
            businessKeyQuery: 'BF-',
            caseSearchResults: {
                page: {
                    totalElements: 2
                },
                '_embedded' : {
                    cases: [{
                        businessKey: 'businessKey1'
                    },{
                        businessKey: 'businessKey2'
                    }]
                }
            }
        };
        const store = configureStore()({
            'case-page': Map({
            }),
            keycloak: props.kc,
            appConfig: props.appConfig
        });

        const wrapper = mount(<Router history={history}>
            <Provider store={store}>
                <CaseResultsPanel {...store} {...props}/>
            </Provider></Router>);

        expect(wrapper).toMatchSnapshot();
    })
});
