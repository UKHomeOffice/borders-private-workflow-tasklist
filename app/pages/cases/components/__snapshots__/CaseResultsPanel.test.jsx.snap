// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseResultsPanel renders search result keys 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "trigs2",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "trigs2",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <withRouter(Connect(Connect(withLog)))
      businessKeyQuery="BF-"
      caseSearchResults={
        Object {
          "_embedded": Object {
            "cases": Array [
              Object {
                "businessKey": "businessKey1",
              },
              Object {
                "businessKey": "businessKey2",
              },
            ],
          },
          "page": Object {
            "totalElements": 2,
          },
        }
      }
      clearActions={[Function]}
      dispatch={[Function]}
      getActions={[Function]}
      getState={[Function]}
      replaceReducer={[Function]}
      searching={false}
      subscribe={[Function]}
    >
      <Connect(Connect(withLog))
        businessKeyQuery="BF-"
        caseSearchResults={
          Object {
            "_embedded": Object {
              "cases": Array [
                Object {
                  "businessKey": "businessKey1",
                },
                Object {
                  "businessKey": "businessKey2",
                },
              ],
            },
            "page": Object {
              "totalElements": 2,
            },
          }
        }
        clearActions={[Function]}
        dispatch={[Function]}
        getActions={[Function]}
        getState={[Function]}
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "trigs2",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "trigs2",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "trigs2",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        replaceReducer={[Function]}
        searching={false}
        subscribe={[Function]}
      >
        <Connect(withLog)
          businessKeyQuery="BF-"
          caseSearchResults={
            Object {
              "_embedded": Object {
                "cases": Array [
                  Object {
                    "businessKey": "businessKey1",
                  },
                  Object {
                    "businessKey": "businessKey2",
                  },
                ],
              },
              "page": Object {
                "totalElements": 2,
              },
            }
          }
          clearActions={[Function]}
          dispatch={[Function]}
          getActions={[Function]}
          getCaseByKey={[Function]}
          getState={[Function]}
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "trigs2",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "trigs2",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "trigs2",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          replaceReducer={[Function]}
          searching={false}
          subscribe={[Function]}
        >
          <withLog
            businessKeyQuery="BF-"
            caseSearchResults={
              Object {
                "_embedded": Object {
                  "cases": Array [
                    Object {
                      "businessKey": "businessKey1",
                    },
                    Object {
                      "businessKey": "businessKey2",
                    },
                  ],
                },
                "page": Object {
                  "totalElements": 2,
                },
              }
            }
            clearActions={[Function]}
            dispatch={[Function]}
            getActions={[Function]}
            getCaseByKey={[Function]}
            getState={[Function]}
            handleError={[Function]}
            handleUnauthorised={[Function]}
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "trigs2",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "trigs2",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "key": "trigs2",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            log={[Function]}
            logFailure={[Function]}
            logSuccess={[Function]}
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            replaceReducer={[Function]}
            resetErrors={[Function]}
            searching={false}
            subscribe={[Function]}
          >
            <CaseResultsPanel
              businessKeyQuery="BF-"
              caseSearchResults={
                Object {
                  "_embedded": Object {
                    "cases": Array [
                      Object {
                        "businessKey": "businessKey1",
                      },
                      Object {
                        "businessKey": "businessKey2",
                      },
                    ],
                  },
                  "page": Object {
                    "totalElements": 2,
                  },
                }
              }
              clearActions={[Function]}
              dispatch={[Function]}
              getActions={[Function]}
              getCaseByKey={[Function]}
              getState={[Function]}
              handleError={[Function]}
              handleUnauthorised={[Function]}
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "canGo": [Function],
                  "createHref": [Function],
                  "entries": Array [
                    Object {
                      "hash": "",
                      "key": "trigs2",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    },
                  ],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "index": 0,
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "trigs2",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "key": "trigs2",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                }
              }
              log={[Function]}
              logFailure={[Function]}
              logSuccess={[Function]}
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
              replaceReducer={[Function]}
              resetErrors={[Function]}
              searching={false}
              subscribe={[Function]}
            >
              <div
                className="govuk-grid-row"
              >
                <div
                  className="govuk-grid-column-one-quarter"
                >
                  <h3
                    className="govuk-heading-m"
                  >
                    Search results
                  </h3>
                  <span
                    className="govuk-caption-m"
                  >
                    Number of cases found
                  </span>
                  <h3
                    className="govuk-heading-m"
                  >
                    2
                  </h3>
                  <ul
                    className="govuk-list"
                  >
                    <li
                      key="businessKey1"
                    >
                      <a
                        className="govuk-link"
                        href=""
                        onClick={[Function]}
                      >
                        businessKey1
                      </a>
                    </li>
                    <li
                      key="businessKey2"
                    >
                      <a
                        className="govuk-link"
                        href=""
                        onClick={[Function]}
                      >
                        businessKey2
                      </a>
                    </li>
                  </ul>
                </div>
                <div
                  className="govuk-grid-column-three-quarters"
                />
              </div>
            </CaseResultsPanel>
          </withLog>
        </Connect(withLog)>
      </Connect(Connect(withLog))>
    </withRouter(Connect(Connect(withLog)))>
  </Provider>
</Router>
`;
